File: algorithms/neetcode/all/increasing-order-search-tree/tree.py
'''
https://www.askpython.com/python/examples/preorder-tree-traversal
https://www.askpython.com/python/examples/inorder-tree-traversal
https://www.askpython.com/python/examples/postorder-tree-traversal-in-python
https://www.askpython.com/python/examples/level-order-binary-tree

'''

from queue import Queue


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


'''
root = TreeNode(5)
root.left = TreeNode(3)
root.left.left = TreeNode(2)
root.left.right = TreeNode(4)
root.left.left.left = TreeNode(1)
root.right = TreeNode(6)
root.right.right = TreeNode(8)
root.right.right.left = TreeNode(7)
root.right.right.right = TreeNode(9)
'''

def insert(root, val):
    if root is None:
        root = TreeNode(val)
        return root
    if val < root.val:
        root.left = insert(root.left, val)
    else:
        root.right = insert(root.right, val)
    return root

root = insert(None,5)
insert(root, 3)
insert(root, 6)
insert(root, 2)
insert(root, 4)
insert(root, 1)
insert(root, 8)
insert(root, 7)
insert(root, 9)



def preorder(root):
    if root == None:
        return
    print(root.val)
    preorder(root.left)
    preorder(root.right)

#preorder(root)

def inorder(root):
    if root == None:
        return
    inorder(root.left)
    print(root.val)
    inorder(root.right)

#inorder(root)

def postorder(root):
    if root == None:
        return
    postorder(root.left)
    postorder(root.right)
    print(root.val)

def levelorder(root):
    if root == None:
        return
    q = Queue()
    q.put(root)
    while not q.empty():
        node = q.get()
        if node == None:
            continue
        print(node.val)
        q.put(node.left)
        q.put(node.right)

if __name__ == '__main__':
    postorder(root)
    print("\n")
    levelorder(root)

File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/blind75/tree/Tree.java
package com.mycompany.leetcode.blind75.tree;

public class Tree {

    // https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/
    public void printInOrder(TreeNode root) {
        if(root == null) {
            return;
        }

        // first recur on left child
        printInOrder(root.left);

        // print the val of node
        System.out.print(root.val + " ");

        // recur on right child
        printInOrder(root.right);
    }

    public void printPreorder(TreeNode root) {
        if(root == null) {
            return;
        }

        // print the val of node
        System.out.print(root.val + " ");

        // first recur on left child
        printPreorder(root.left);

        // recur on right child
        printPreorder(root.right);
    }


    public void printPostorder(TreeNode root) {
        if(root == null) {
            return;
        }


        // first recur on left child
        printPostorder(root.left);

        // recur on right child
        printPostorder(root.right);

        // print the val of node
        System.out.print(root.val + " ");
    }


    // Level Order Traversal (Naive approach):
    // https://www.geeksforgeeks.org/level-order-tree-traversal/
    // Function to print level order traversal of tree
    public void printLevelOrder(TreeNode root)
    {
        int h = height(root);
        int i;
        for (i = 1; i <= h; i++)
            printCurrentLevel(root, i);
    }

    // Compute the "height" of a tree -- the number of
    // nodes along the longest path from the root node
    // down to the farthest leaf node.
    private int height(TreeNode root)
    {
        if (root == null)
            return 0;
        else {

            // Compute  height of each subtree
            int lheight = height(root.left);
            int rheight = height(root.right);

            // use the larger one
            if (lheight > rheight)
                return (lheight + 1);
            else
                return (rheight + 1);
        }
    }

    // Print nodes at the current level
    private void printCurrentLevel(TreeNode root, int level)
    {
        if (root == null)
            return;
        if (level == 1)
            System.out.print(root.val + " ");
        else if (level > 1) {
            printCurrentLevel(root.left, level - 1);
            printCurrentLevel(root.right, level - 1);
        }
    }
}

// https://www.javatpoint.com/tree-traversal
// https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/
